system_zero/src/permutation_unit.rs

```diff
@@ -2,36 +2,117 @@ use plonky2::field::extension_field::{Extendable, FieldExtension};
+use plonky2::hash::poseidon::{HALF_N_FULL_ROUNDS, N_PARTIAL_ROUNDS};
 use plonky2::plonk::circuit_builder::CircuitBuilder;
 use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
 use starky::vars::StarkEvaluationTargets;
 use starky::vars::StarkEvaluationVars;
 
-use crate::column_layout::{col_permutation_input, col_permutation_output, NUM_COLUMNS};
+use crate::column_layout::{
+    col_permutation_full_first_after_mds, col_permutation_full_first_mid_sbox,
+    col_permutation_full_second_after_mds, col_permutation_full_second_mid_sbox,
+    col_permutation_input, col_permutation_partial_after_sbox, col_permutation_partial_mid_sbox,
+    NUM_COLUMNS,
+};
 use crate::public_input_layout::NUM_PUBLIC_INPUTS;
 use crate::system_zero::SystemZero;
 
+fn constant_layer<F, FE, P, const D2: usize>(
+    mut state: [P; SPONGE_WIDTH],
+    round: usize,
+) -> [P; SPONGE_WIDTH]
+where
+    F: RichField,
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    // One day I might actually vectorize this, but today is not that day.
+    for i in 0..P::WIDTH {
+        let mut unpacked_state = [P::Scalar::default(); SPONGE_WIDTH];
+        for j in 0..SPONGE_WIDTH {
+            unpacked_state[j] = state[j].as_slice()[i];
+        }
+        F::constant_layer_field(&mut unpacked_state, round);
+        for j in 0..SPONGE_WIDTH {
+            state[j].as_slice_mut()[i] = unpacked_state[j];
+        }
+    }
+    state
+}
+
+fn mds_layer<F, FE, P, const D2: usize>(mut state: [P; SPONGE_WIDTH]) -> [P; SPONGE_WIDTH]
+where
+    F: RichField,
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    for i in 0..P::WIDTH {
+        let mut unpacked_state = [P::Scalar::default(); SPONGE_WIDTH];
+        for j in 0..SPONGE_WIDTH {
+            unpacked_state[j] = state[j].as_slice()[i];
+        }
+        unpacked_state = F::mds_layer_field(&unpacked_state);
+        for j in 0..SPONGE_WIDTH {
+            state[j].as_slice_mut()[i] = unpacked_state[j];
+        }
+    }
+    state
+}
+
 impl<F: RichField + Extendable<D>, const D: usize> SystemZero<F, D> {
-    pub(crate) fn generate_permutation_unit(&self, values: &mut [F; NUM_COLUMNS]) {
+    pub(crate) fn generate_permutation_unit(values: &mut [F; NUM_COLUMNS]) {
         // Load inputs.
-        let mut state = [F::ZERO; SPONGE_WIDTH];
+        let mut state = [F::ZEROS; SPONGE_WIDTH];
         for i in 0..SPONGE_WIDTH {
             state[i] = values[col_permutation_input(i)];
         }
 
-        // TODO: First full rounds.
-        // TODO: Partial rounds.
-        // TODO: Second full rounds.
+        for r in 0..HALF_N_FULL_ROUNDS {
+            F::constant_layer(&mut state, r);
 
-        // Write outputs.
-        for i in 0..SPONGE_WIDTH {
-            values[col_permutation_output(i)] = state[i];
+            for i in 0..SPONGE_WIDTH {
+                let state_cubed = state[i] * state[i].square();
```

Comment:
nit: could use `cube()`

Reply:
I haven't added `cube` to packed fields, because I'm not convinced I can do it faster than a square and a mul. But that also makes it trivial to implement. Do you think it's worth adding?